athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsMetadataHandler.java

```diff
@@ -0,0 +1,320 @@
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.amazonaws.athena.connector.lambda.metadata.ListTablesRequest.UNLIMITED_PAGE_SIZE_VALUE;
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.CLASSIFICATION_GLUE_TABLE_PARAM;
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.GCS_CREDENTIAL_KEYS_ENV_VAR;
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.GCS_SECRET_KEY_ENV_VAR;
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.STORAGE_SPLIT_JSON;
+import static com.amazonaws.athena.connectors.gcs.GcsSchemaUtils.buildTableSchema;
+import static com.amazonaws.athena.connectors.gcs.GcsUtil.getGcsCredentialJsonString;
+import static com.amazonaws.athena.connectors.gcs.GcsUtil.splitAsJson;
+import static com.amazonaws.athena.connectors.gcs.storage.datasource.StorageDatasourceFactory.createDatasource;
+
+public class GcsMetadataHandler
+        extends GlueMetadataHandler
+{
+    public static final String TABLE_FILTER_IDENTIFIER = "gs://";
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcsMetadataHandler.class);
+    /**
+     * used to aid in debugging. Athena will use this name in conjunction with your catalog id
+     * to correlate relevant query errors.
+     */
+    private static final String SOURCE_TYPE = "gcs";
+    // Env variable name used to indicate that we want to disable the use of Glue DataCatalog for supplemental
+    // metadata and instead rely solely on the connector's schema inference capabilities.
+    private static final String GLUE_ENV = "disable_glue";
+    private static final CharSequence GCS_FLAG = "gcs";
+    private static final DatabaseFilter DB_FILTER = (Database database) -> (database.getLocationUri() != null && database.getLocationUri().contains(GCS_FLAG));
+    // used to filter out Glue tables which lack indications of being used for DDB.
+    private static final TableFilter TABLE_FILTER = (Table table) -> table.getStorageDescriptor().getLocation().contains(TABLE_FILTER_IDENTIFIER)
+            || (table.getParameters() != null && GcsUtil.isSupportedFileType(table.getParameters().get("classification")))
+            || (table.getStorageDescriptor().getParameters() != null && GcsUtil.isSupportedFileType(table.getStorageDescriptor().getParameters().get("classification")));
+    private final StorageMetadata datasource;
+    private final AWSGlue glueClient;
+
+    public GcsMetadataHandler() throws IOException
+    {
+        super((System.getenv(GLUE_ENV) != null && !"false".equalsIgnoreCase(System.getenv(GLUE_ENV))), SOURCE_TYPE);
+        String gcsCredentialsJsonString = getGcsCredentialJsonString(this.getSecret(System.getenv(GCS_SECRET_KEY_ENV_VAR)), GCS_CREDENTIAL_KEYS_ENV_VAR);
+        this.datasource = createDatasource(gcsCredentialsJsonString, System.getenv());
+        this.glueClient = getAwsGlue();
```

Comment:
Since majority of metadata handler method require glue client to work properly. can we add requireNonNull for getAwsGlue)? in case customer using wrong flag to disable glue

Reply:
added requireNotNull in the constructor.